File: problems/1443/solution.cpp
class Solution {
public:
    long dfs(int prev, int curr, vector<int> m[], vector<bool>& hasApple) {
        long ans = 0 ; 

        for (int child : m[curr]) {
            if (child != prev) {
                long cnt = dfs(curr, child, m, hasApple);

                if (cnt > 0 || hasApple[child]) {
                    ans += (cnt + 2);
                }
            }
        }

        return ans;
    }
    
    int minTime(int n, vector<vector<int>>& edges, vector<bool>& hasApple) {
        vector<int> m[n];
        
        for (const vector<int>& edge: edges) {
            m[edge[0]].push_back(edge[1]);
            m[edge[1]].push_back(edge[0]);
        }

        return dfs(-1, 0, m, hasApple);
    }
};

File: problems/1443/solution.py
class Solution:
    def minTime(self, n: int, edges: List[List[int]], hasApple: List[bool]) -> int:
        m = {}
        
        for f, t in edges:
            if f not in m:
                m[f] = set()
            
            m[f].add(t)
        
        def go(vertex):
            depth = 0
            
            if hasApple[vertex] and vertex:
                depth += 1
            
            if vertex in m:
                childsDepth = 0
                
                for child in m[vertex]:
                    childsDepth += go(child)
                
                if not depth and childsDepth and vertex:
                    depth += 1
                
                depth += childsDepth
            
            return depth
        
        return go(0) * 2
            